### PR TITLE
Add Microgateway performance test scripts

### PR DESCRIPTION
**Pull request by tharikaGitHub**
## Purpose
This PR adds performance test components for API Microgateway.

### CODE REVIEW

File: components/jwtgenerator/pom.xml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    <parent>
+        <groupId>org.wso2.performance.apim</groupId>
+        <artifactId>performance-apim</artifactId>
+        <version>0.1.1-SNAPSHOT</version>
+        <relativePath>../../pom.xml</relativePath>
+    </parent>
+
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>org.wso2.am.microgw</groupId>
```
Comment (chrishantha) [inline components/jwtgenerator/pom.xml]: Do not specify group ID

File: components/jwtgenerator/pom.xml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    <artifactId>org.wso2.am.microgw.jwt-generator</artifactId>
```
Comment (chrishantha) [inline components/jwtgenerator/pom.xml]: Use the project group ID "org.wso2.performance.apim" as a prefix

File: components/jwtgenerator/pom.xml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    <version>1.0.0</version>
+
+    <dependencies>
+        <dependency>
```
Comment (chrishantha) [inline components/jwtgenerator/pom.xml]: Use dependency management in parent pom and use properties for version.

File: components/jwtgenerator/pom.xml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                    <descriptorRefs>
+                        <descriptorRef>jar-with-dependencies</descriptorRef>
+                    </descriptorRefs>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+
+    <properties>
+        <main.class>org.wso2.am.microgw.jwt.JWTGenerator</main.class>
```
Comment (chrishantha) [inline components/jwtgenerator/pom.xml]: This is not required.

File: components/jwtgenerator/pom.xml [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    </properties>
+</project>
```
Comment (chrishantha) [inline components/jwtgenerator/pom.xml]: Add new line at the end of file.

File: components/jwtgenerator/src/main/java/org/wso2/am/microgw/jwt/JWTGenerator.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+import java.security.KeyStore;
+import java.security.PrivateKey;
+import java.security.Signature;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Base test class for CLI based tests
```
Comment (chrishantha) [inline components/jwtgenerator/src/main/java/org/wso2/am/microgw/jwt/JWTGenerator.java]: Is this correct?

File: components/jwtgenerator/src/main/java/org/wso2/am/microgw/jwt/JWTGenerator.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.am.microgw.jwt;
+
+import org.apache.commons.logging.Log;
```
Comment (chrishantha) [inline components/jwtgenerator/src/main/java/org/wso2/am/microgw/jwt/JWTGenerator.java]: Use slf4j

File: components/jwtgenerator/src/main/java/org/wso2/am/microgw/jwt/JWTGenerator.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    public static final int VALIDITY_PERIOD = 3600 * 24 * 365;
+
+    public static void main(String[] args) throws Exception {
+        JWTGenerator jwtGenerator = new JWTGenerator();
+        ApplicationDTO application = new ApplicationDTO();
+
+        String apiName, context, version, appName, appTier, subsTier;
+        int appId;
+        int i = 0;
+        apiName = args[i++];
```
Comment (chrishantha) [inline components/jwtgenerator/src/main/java/org/wso2/am/microgw/jwt/JWTGenerator.java]: Use jcommander. For example, see https://github.com/wso2/performance-common/blob/4430a2808d342e7995b9c0e9dedfd17a2d60a7d5/components/netty-http-echo-service/src/main/java/org/wso2/performance/common/netty/echo/EchoHttpServer.java#L48-L49

File: distribution/scripts/apim/micro-gw/create-apis.sh [hunk 9]
```diff
@@ -0,0 +1,370 @@
+#!/bin/bash
```
Comment (chrishantha) [inline distribution/scripts/apim/micro-gw/create-apis.sh]: Why did you create this file?

File: distribution/scripts/apim/micro-gw/create-micro-gw.sh [hunk 10]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/expect
```
Comment (chrishantha) [inline distribution/scripts/apim/micro-gw/create-micro-gw.sh:1]: Add license header

File: distribution/scripts/apim/micro-gw/micro-gw-start.sh [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+export JAVA_HOME="${jvm_dir}"
+
+echo "Starting Microgateway"
+pushd /home/ubuntu/micro-gw-${label}/bin/
+bash gateway > /dev/null &
+popd
+
+echo "Waiting for Microgateway to start"
+
+if [ -e "/home/ubuntu/micro-gw-${label}/bin/gateway.pid" ]; then
```
Comment (chrishantha) [inline distribution/scripts/apim/micro-gw/micro-gw-start.sh]: Shouldn't we use a loop here?

File: distribution/scripts/cloudformation/templates/apim_perf_test_cfn.yaml [hunk 12]
```diff
@@ -1,4 +1,4 @@
-{#
+#{
```
Comment (chrishantha) [inline distribution/scripts/cloudformation/templates/apim_perf_test_cfn.yaml]: Why did you change this?

File: pom.xml [hunk 13]
```diff
@@ -58,6 +58,12 @@
                 <version>${performance.common.version}</version>
                 <type>tar.gz</type>
             </dependency>
+            <dependency>
```
Comment (chrishantha) [inline pom.xml:72]: Why did you add this?

**Review by chrishantha — CHANGES_REQUESTED** [review]

File: distribution/scripts/apim/micro-gw/create-apis.sh (hunk 9 above)
Comment (tharikaGitHub) [reply distribution/scripts/apim/micro-gw/create-apis.sh]: This file is different from apim/create-apis.sh. This does not create a mediation API. Will remove the sections not required for Microgateway testing.

File: distribution/bin.xml [hunk 14]
```diff
@@ -41,5 +41,17 @@
             <filtered>true</filtered>
         </fileSet>
     </fileSets>
+    <files>
+        <file>
```
Comment (chrishantha) [inline distribution/bin.xml:52]: Use dependencySet. Look at performance-common. https://github.com/wso2/performance-common/blob/master/distribution/bin.xml#L32-L57

**Review by chrishantha — CHANGES_REQUESTED** [review]

File: distribution/pom.xml [hunk 15]
```diff
@@ -35,6 +35,12 @@
             <artifactId>performance-common-distribution</artifactId>
             <type>tar.gz</type>
         </dependency>
+        <dependency>
+            <groupId>org.wso2.performance.apim</groupId>
+            <artifactId>jwt-generator</artifactId>
+            <version>1.0.0</version>
```
Comment (chrishantha) [inline distribution/pom.xml]: Do not use version here.

**Review by chrishantha — CHANGES_REQUESTED** [review]

File: distribution/pom.xml [hunk 16]
```diff
@@ -35,6 +35,12 @@
             <artifactId>performance-common-distribution</artifactId>
             <type>tar.gz</type>
         </dependency>
+        <dependency>
+            <groupId>org.wso2.performance.apim</groupId>
+            <artifactId>jwt-generator</artifactId>
+            <version>${jwt.generator.version}</version>
```
Comment (chrishantha) [inline distribution/pom.xml]: Version shouldn't be here.

File: distribution/scripts/apim/micro-gw/generate-jwt-tokens.sh [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+if [[ -f $tokens_file ]]; then
+    mv $tokens_file /tmp
+fi
+
+echo "Generating Tokens.........."
+
+for (( c=1; c <= $tokens_count; c++ ))
+do
+	JWT_TOKEN=$(java -jar /home/ubuntu/jwt-generator/jwt-generator-1.0.0.jar --api-name "echo" --context "/echo" --version "1.0.0" --app-name "DefaultApplication" --app-tier "Unlimited" --subs-tier "Unlimited" --app-id 1)
```
Comment (chrishantha) [inline distribution/scripts/apim/micro-gw/generate-jwt-tokens.sh]: Do not hard code path and version here.

File: distribution/scripts/apim/micro-gw/micro-gw-start.sh [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+    esac
+done
+shift "$((OPTIND - 1))"
+
+if [[ -z $heap_size ]]; then
+    echo "Please provide the heap size for the API Microgateway."
+    exit 1
+fi
+
+label="echo-mgw"
```
Comment (chrishantha) [inline distribution/scripts/apim/micro-gw/micro-gw-start.sh]: This should be an option.

File: distribution/scripts/apim/micro-gw/micro-gw-start.sh [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+
+jvm_dir=""
+for dir in /usr/lib/jvm/jdk1.8*; do
+    [ -d "${dir}" ] && jvm_dir="${dir}" && break
+done
+export JAVA_HOME="${jvm_dir}"
+
+echo "Starting Microgateway"
+pushd /home/ubuntu/micro-gw-${label}/bin/
+bash gateway > /dev/null &
```
Comment (chrishantha) [inline distribution/scripts/apim/micro-gw/micro-gw-start.sh:108]: This is not recommended. Execute as ./gateway

File: pom.xml [hunk 20]
```diff
@@ -149,9 +187,13 @@
     </profiles>
 
     <properties>
+        <jcommander.version>1.72</jcommander.version>
+        <slf4j.version.range>[1.7.0,1.8.0)</slf4j.version.range>
+        <org.json.version>20180130</org.json.version>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <performance.apim.version>0.1.1-SNAPSHOT</performance.apim.version>
         <performance.common.version>0.2.1-SNAPSHOT</performance.common.version>
+        <jwt.generator.version>1.0.0</jwt.generator.version>
```
Comment (chrishantha) [inline pom.xml]: Remove jwt.generator.version. We don't need a specific version for jwt-generator

File: distribution/scripts/cloudformation/run-micro-gw-performance-tests.sh [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+    echo "Throughput (Requests/sec)"
+    echo "Average Response Time (ms)"
+    echo "Standard Deviation of Response Time (ms)"
+    echo "99th Percentile of Response Time (ms)"
+    echo "WSO2 API Microgateway GC Throughput (%)"
+    echo "Average WSO2 API Microgateway Memory Footprint After Full GC (M)"
+}
+export -f get_columns
+
+export aws_cloudformation_template_filename="apim_micro_gw_perf_test_cfn.yaml"
```
Comment (chrishantha) [inline distribution/scripts/cloudformation/run-micro-gw-performance-tests.sh:128]: also export run_performance_tests_script_name="run-micro-gw-performance-tests.sh"

File: components/jwt-generator/src/main/java/org/wso2/am/microgw/jwt/JWTGenerator.java [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+/**
+ * JWT Token Generator for Microgateway Performance Tests
+ */
+public class JWTGenerator {
+
+    private static final String KEY_TYPE_PRODUCTION = "PRODUCTION";
+    private static final String WSO2CARBON = "wso2carbon";
+    private static final int VALIDITY_PERIOD = 3600 * 24 * 365;
+
+    @Parameter(names = "--api-name", description = "API Name") private String apiName;
```
Comment (chrishantha) [inline components/jwt-generator/src/main/java/org/wso2/am/microgw/jwt/JWTGenerator.java]: Did you format code?

File: distribution/scripts/apim/micro-gw/generate-jwt-tokens.sh (hunk 17 above)
Comment (chrishantha) [reply distribution/scripts/apim/micro-gw/generate-jwt-tokens.sh]: https://github.com/chrishantha/performance-common/blob/master/distribution/scripts/jtl-splitter/jtl-splitter.sh#L58

**Review by chrishantha — CHANGES_REQUESTED** [review]

File: distribution/scripts/apim/micro-gw/micro-gw-start.sh (hunk 19 above)
Comment (tharikaGitHub) [reply distribution/scripts/apim/micro-gw/micro-gw-start.sh:108]: Kept as bash gateway due to a limitation in the script.

File: pom.xml [hunk 23]
```diff
@@ -149,6 +193,9 @@
     </profiles>
 
     <properties>
+        <jcommander.version>1.72</jcommander.version>
+        <slf4j.version.range>[1.7.0,1.8.0)</slf4j.version.range>
```
Comment (chrishantha) [inline pom.xml]: This is not required.

File: pom.xml [hunk 24]
```diff
@@ -149,6 +193,9 @@
...
+        <org.json.version>20180130</org.json.version>
```
Comment (chrishantha) [inline pom.xml]: There is a newer version: https://search.maven.org/artifact/org.json/json/20180813/bundle

File: distribution/scripts/cloudformation/run-micro-gw-performance-tests.sh [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+
+    export wso2am_distribution_filename=$(basename $wso2am_distribution)
+
+    if [[ ${wso2am_distribution_filename: -4} != ".zip" ]]; then
+        echo "WSO2 API Manager distribution must have .zip extension"
+        exit 1
+    fi
+
+    if [[ ! -f $wso2am_micro_gw_distribution ]]; then
+            echo "Please provide WSO2 API Microgateway distribution."
```
Comment (chrishantha) [inline distribution/scripts/cloudformation/run-micro-gw-performance-tests.sh:74]: Format code: Use VSCode

**Review by chrishantha — CHANGES_REQUESTED** [review]

**Review by chrishantha — APPROVED** [review]

**Pull request merged**